File: solutions/3880-minimum-time-to-visit-all-houses/minimum-time-to-visit-all-houses.java
class Solution {
    public long minTotalTime(int[] forward, int[] backward, int[] queries) {
        int n = forward.length;
        long[] fwd = new long[n+1], bwd = new long[n+1];
        for (int i = 1; i <= n; i++) {
            fwd[i] = fwd[i-1] + forward[i-1];
            bwd[i] = bwd[i-1] + (i < n ? backward[i] : backward[0]);
        }
        long res = 0;
        int cur = 0;
        for (int q : queries) {
            int u = cur, v = q;
            long cw = u <= v
                ? fwd[v] - fwd[u]
                : fwd[n] - (fwd[u] - fwd[v]);
            long ccw = u >= v
                ? bwd[u] - bwd[v]
                : bwd[n] - (bwd[v] - bwd[u]);
            res += Math.min(cw, ccw);
            cur = v;
        }
        return res;
    }
}

File: solutions/3880-minimum-time-to-visit-all-houses/minimum-time-to-visit-all-houses.js
/**
 * @param {number[]} forward
 * @param {number[]} backward
 * @param {number[]} queries
 * @return {number}
 */
var minTotalTime = function(forward, backward, queries) {
    const n = forward.length;
    const fwd = new Float64Array(n+1);
    const bwd = new Float64Array(n+1);
    for (let i = 1; i <= n; i++) {
        fwd[i] = fwd[i-1] + forward[i-1];
    }
    for (let i = 1; i < n; i++) {
        bwd[i] = bwd[i-1] + backward[i];
    }
    bwd[n] = bwd[n-1] + backward[0];
    const totalFwd = fwd[n], totalBwd = bwd[n];

    let ans = 0;
    let cur = 0;
    const m = queries.length;
    for (let i = 0; i < m; i++) {
        const v = queries[i];
        const u = cur;
        // clockwise distance
        let cw = u <= v
            ? fwd[v] - fwd[u]
            : totalFwd - (fwd[u] - fwd[v]);
        // counter‐clockwise distance
        let ccw = u >= v
            ? bwd[u] - bwd[v]
            : totalBwd - (bwd[v] - bwd[u]);
        ans += (cw < ccw ? cw : ccw);
        cur = v;
    }
    return ans;
};